<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="yolo_model" default="yolov8n.pt"/>
  <arg name="publish_rate" default="10"/>
  <arg name="detection_topic" default="detection_result"/>
  <arg name="input_topic" default="kitti/camera_color_left/image_raw"/>
  <arg name="conf_thres" default="0.25"/>
  <arg name="iou_thres" default="0.45"/>
  <arg name="max_det" default="300"/>
  <arg name="tracker" default="bytetrack.yaml"/>
  <arg name="debug" default="true"/>
  <arg name="debug_conf" default="true"/>
  <arg name="debug_line_width" default="1"/>
  <arg name="debug_font_size" default="1"/>
  <arg name="debug_font" default="Arial.ttf"/>
  <arg name="debug_labels" default="true"/>
  <arg name="debug_boxes" default="true"/>
  <arg name="camera_info_topic" default="kitti/camera_color_left/camera_info"/>
  <arg name="lidar_topic" default="kitti/velo/pointcloud"/>
  <arg name="detection3d_topic" default="detection3d_result"/>
  <arg name="cluster_tolerance" default="0.5"/>
  <arg name="min_cluster_size" default="100"/>
  <arg name="max_cluster_size" default="25000"/>
  <!-- Object detection node -->
  <node name="tracker_node" pkg="ultralytics_ros" type="tracker_node.py" output="screen">
    <param name="yolo_model" value="$(arg yolo_model)"/>
    <param name="publish_rate" value="$(arg publish_rate)"/>
    <param name="detection_topic" value="$(arg detection_topic)"/>
    <param name="input_topic" value="$(arg input_topic)"/>
    <param name="conf_thres" value="$(arg conf_thres)"/>
    <param name="iou_thres" value="$(arg iou_thres)"/>
    <param name="max_det" value="$(arg max_det)"/>
    <param name="tracker" value="$(arg tracker)"/>
    <param name="debug" value="$(arg debug)"/>
    <param name="debug_conf" value="$(arg debug_conf)"/>
    <param name="debug_line_width" value="$(arg debug_line_width)"/>
    <param name="debug_font_size" value="$(arg debug_font_size)"/>
    <param name="debug_font" value="$(arg debug_font)"/>
    <param name="debug_labels" value="$(arg debug_labels)"/>
    <param name="debug_boxes" value="$(arg debug_boxes)"/>
    <rosparam>
      # classes: [0] (person)
      # classes: [1] (bicycle)
      # classes: [2] (car)
      # ...
      # See also : https://github.com/ultralytics/ultralytics/blob/main/ultralytics/datasets/coco128.yaml
    </rosparam>
  </node>
  <!-- Object detection with pointcloud node -->
  <node name="tracker_with_cloud_node" pkg="ultralytics_ros" type="tracker_with_cloud_node" output="screen">
    <param name="camera_info_topic" value="$(arg camera_info_topic)"/>
    <param name="lidar_topic" value="$(arg lidar_topic)"/>
    <param name="detection2d_topic" value="$(arg detection_topic)"/>
    <param name="detection3d_topic" value="$(arg detection3d_topic)"/>
    <param name="cluster_tolerance" value="$(arg cluster_tolerance)"/>
    <param name="min_cluster_size" value="$(arg min_cluster_size)"/>
    <param name="max_cluster_size" value="$(arg max_cluster_size)"/>
  </node>
  <!-- rviz -->
  <node if="$(arg debug)" pkg="rviz" type="rviz" name="rviz" args="-d $(find ultralytics_ros)/rviz/kitti.rviz"/>
</launch>
